app/scripts/lib/AbstractPetnamesBridge.ts

```diff
@@ -0,0 +1,240 @@
+import {
+  NameController,
+  NameStateChange,
+  NameType,
+  SetNameRequest,
+} from '@metamask/name-controller';
+import { RestrictedControllerMessenger } from '@metamask/base-controller';
+
+// Use the same type for both the source entries and the argument to NameController::setName.
+export type PetnameEntry = SetNameRequest;
+
+// The type of change that occurred.
+export enum ChangeType {
+  ADDED = 'ADDED',
+  UPDATED = 'UPDATED',
+  DELETED = 'DELETED',
+}
+
+type SyncDirection = 'Source->Petnames' | 'Petnames->Source';
+
+// A list of changes, grouped by type.
+type ChangeList = Record<ChangeType, PetnameEntry[]>;
+
+type AllowedEvents = NameStateChange;
+
+export type PetnamesBridgeMessenger = RestrictedControllerMessenger<
+  'PetnamesBridge',
+  never,
+  AllowedEvents,
+  never,
+  AllowedEvents['type']
+>;
+
+/**
+ * Abstract class representing a bridge between petnames and a data source.
+ * Provides methods for synchronizing petnames with the data source and handling changes.
+ */
+export abstract class AbstractPetnamesBridge {
+  #isTwoWay: boolean;
+
+  #nameController: NameController;
+
+  #synchronizingDirection: SyncDirection | 'NONE' = 'NONE';
```

Comment:
Rather than having a `NONE` entry, is it simpler to set it to `null` or `undefined` since the platform provides that state for us?

Reply:
Done in next update.